starting RStudio content (#22)
fhdsl/Tools_for_Reproducible_Workflows_in_R: 04-rstudio-tour.Rmd

# A Tour of RStudio

In this chapter we will talk about a very useful R-related tool called RStudio. RStudio is an environment for using R that can be extremely helpful for writing code and making your analyses reproducible. 

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "A Tour of RStudio, Learning Objectives are to be able to, Recognize why RStudio is useful, Install RStudio and get started with it, Navigate RStudio"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g21a84b32106_0_0")
```

## Why use RStudio?

RStudio is what is called an **integrated development environment (IDE)** for writing code in R (although it also has compatibility for other languages).

It is designed to make working in R easier in a variety of ways:

> - it helps you write code by making suggestions to complete what you have written - currently this is mostly for suggesting package names or functions (which are specific pieces of code that accomplish a particular task, often packages have several functions)

Reviewer: You could start this list off with `it helps you:` so you don't have to repeat this phrase everytime.

Author: fixed!

- it helps you view the output of your code, this is especially true for creating reports or viewing plots
- it helps you find errors in your code
- it helps you keep track of any objects that you have assigned in R
- it helps you orient yourself in terms of the files on your computer 
- it can help track changes in your code and other files over time

<div class = "dictionary">

- **IDE** - Integrated Development Environment - a computing environment for writing code, debugging code, and looking at the output of your code 

> - **RStudio** - an IDE designed especially for writing R code

Reviewer: Have we defined a package yet?

Author: yes we did :)

Author: maybe we should have dictionary boxes remind people? what do you think?

- **function** - a specific piece of code that performs a task - packages in R often have several functions
- **objects** - objects in R could be anything that you can refer to with some name to recall again such as a data tables, vectors, functions, plots and more. 

</div>

We will dive deeper into these benefits later once we get started with RStudio, but first we will discuss how to make sure you have it downloaded and installed on your computer.

## Installing RStudio

> In case you don't yet have RStudio on your computer, we will walk you through the process of getting started.

Reviewer: Probably don't need this sentence in its own separate section.

Author: Oh I just really dislike headers with nothing but other headers right below :P

Author: anything else we could add? so it is not just one sentence

Reviewer: Could probably delete the header too.

### Installing and Updating R

You first need to make sure that you have R. R is not the same as RStudio. R is instead the libraries needed to use R code on your computer and it is needed so that you can use RStudio. It is also a good idea to update the version of R that you are using periodically. 

<details> <summary> Click here for directions if you have never installed R before on your computer. </summary> 

> You can install the latest version of R from the R project site located here: https://www.r-project.org/

Reviewer: I think if you install RStudio, you don't need to install R separately but I could be wrong.

Author: I think that you do - according to this source: https://rstudio-education.github.io/hopr/starting.html

Reviewer: Sounds good!

From here you can click on the menu option that says CRAN on the far left to start. Recall that CRAN stands for the **Comprehensive R Archive Network**.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Once you go to the r project website, you can click on the CRAN button to download the latest version of R"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_0")
```

> This will take you to a website with a list of what are called [mirrors](https://cran.r-project.org/mirrors.html), which are locations that have the same exact copy of R but are dispersed geographically mostly to improve download speeds for users. Nothing bad will happen if you click on a mirror that isn't closest to you, but it can improve download speeds overall if everyone tries to. 

Reviewer: ```suggestion
This will take you to a website with a list of what are called [mirrors](https://cran.r-project.org/mirrors.html), which are locations that have the same exact copy of R but are dispersed geographically mostly to improve download speeds for users. Nothing bad will happen if you click on a mirror that isn't closest to you, but it can improve download speeds for everyone overall. 
```

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "CRAN mirror list"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_11")
```

Once you click on one of the mirror links you will be taken to a new page to download R. For example, you could click on the Iowa state University mirror if you are located in the US somewhere.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Page to download R for computers with different operating systems, such as Mac or Windows"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_19")
```

You would want to click on the appropriate link for your computer. For example, if you have a Windows machine, click the link for Windows. 

This will take you to a new page to select the appropriate link to download R. For Mac users this might be the most recent version of R which will look like R and several numbers afterwards. 

</details> 

<details><summary>Click here for instructions on how to update R.</summary>

> To update R, if you are using a Mac or Linux computer, you can follow the directions of installing R the first time.

Reviewer: Is there something we can add here for someone to know how to check if they need to update R before they do so? I believe RStudio sends notifications now when you open it to let you know you should update it. 

Here's a screenshot you can use if you like: 

<img width="259" alt="Screen Shot 2023-03-30 at 8 45 17 AM" src="https://user-images.githubusercontent.com/23458084/228840156-3c7c5e92-686e-40f8-a7f1-2338d349ca05.png">

Author: The method I showed will say that it is up-to-date if you do not need to update it. I can add a screen shot about that

Author: done!

> If you have a Windows computer, you can use the following code to update your version of R within an R session. You can start an R session by typing `R` into whatever command-line shell like Windows Terminal that you might have been using in the past. Then copy and paste this code and press enter. 

Reviewer: ```suggestion
If you have a Windows computer, you can use the following code to update your version of R within an R session. You can start an R session by typing `R` into Command Prompt window. If you have not used the Command Prompt window, [read instructions here](https://www.makeuseof.com/tag/a-beginners-guide-to-the-windows-command-line/) about how to find it. After opening your Command Prompt window, copy and paste this code and press enter. 
```

```{r, eval=FALSE}
# Check for the install r package and install if needed
if(!require(installr)) {
  install.packages("installr"); 
  require(installr) #load installr pakage
} 

updateR()#update your version of R
```

</details>

### Installing RStudio

> Next we want to download and install RStudio. You can do so by going to the Posit website at this link: https://posit.co/.

Reviewer: ```suggestion
Next we want to download and install RStudio. You can do so by going to the Posit website at this link: https://posit.co/. Note that you should be able to download the completely free option. 
```

Reviewer: Honestly you could just take advantage of the instructions for installing R that RStudio has. That might be easier.

<div class = "notice">

> Posit is a software company that used to be called RStudio that develops open-source R tools and packages. It is a Public Benefit Corporation (PBC) and a Certified B Corporation®, so it is committed to creating software that benefits the public. See [here](https://posit.co/about/pbc-report/) for more information.

Reviewer: ```suggestion
Posit is a software company that used to be called RStudio that develops open-source data science tools and packages. It is a Public Benefit Corporation (PBC) and a Certified B Corporation®, so it is committed to creating software that benefits the public. See [here](https://posit.co/about/pbc-report/) for more information.
```

Technically they do stuff other than R now, which is why they switched their name away from RStudio.

Author: yes, good catch - I just was focusing on R in my thoughts for this audience

</div>

<details><summary> Click here for instructions on how to download and install RStudio. </summary>

> ```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "The Posit website for downloading RStudio"}

Reviewer: We might want to be careful about including screenshots of their website since it will be a maintenance issue to keep up with them everytime they change their website. I'm not sure if they upkeep for this is worth the extra instructions for installation. Posit usually does a decent job at explaining how to install their software.

Author: I have a disclaimer somewhere that it might look slightly different - should we add more? (since I already created this)

Reviewer: Nah. That's probably good enough.

ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_30")
```
Note that the website may look slightly different when you visit it.

There should be a download button on the upper right corner. This will take you to another page to choose if you want the free or paid version of RStudio. The free version should be enough for most users.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "The free version of RStudio is likely all you need."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_37")
```

Then you need to scroll down to select the appropriate download for your computer based on what kind of computer you have.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Select the appropriate link to download RStudio according to the type of computer you have."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_37")
```

> Note that by the time you read this the versions will likely have changed and there may be slight variations in how the website appears.

Reviewer: Lol. I see your strategy. Well, IDK we can figure out what makes sense here.

You should then be directed by your computer on how to install RStudio once the download is complete. You may need to go to your downloads first and click on the RStudio file that was downloaded to start this process.

For Mac users, note that you will need to move the RStudio icon into the icon that looks like the Applications folder. 

Drag and drop RStudio into the Applications folder to install on a Mac

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Drag and drop RStudio into the Applications folder to install on a Mac."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_58")
```

See[here](https://jhudatascience.org/intro_to_r/modules/Setup/R_and_RStudio_installation.html) for more information on the process of installing RStudio.

If you run into trouble, check the following:

> - did you install the correct version of software for your operating system?

Reviewer: ```suggestion
- Did you install the correct version of software for your operating system?
```

> - did you install the correct version of software for your operating system?

Reviewer: I find it unlikely this will be the troubleshooting solution for many people but I could be wrong.

It is also a good idea to keep RStudio up-to-date. New features become available as the Posit team works on developing RStudio. So if you already have RStudio, you might want to check to see if your version is up-to-date.

To check for updates you can go to the `Help` menu at the top of RStudio and then click on `Check for Updates`. 

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Check for updates for RStudio in the Help menu of RStudio."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_50")
```

## Navigating RStudio

Now that you hopefully have RStudio running on your machine, we will walk you through some of the major features that can really help you with your data analyses.

### Default Layout

First it is important to be familiar with the layout. When you first open RStudio you will see 3 panes.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "When RStudio is first opened you will see 3 panes"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_67")
```

> <details><summary>If your RStudio looks different click here.</summary>

Reviewer: ```suggestion
<details><summary>If you'd like to change your RStudio layout click here</summary>
```
If it looks different, its because they've already changed their layout and are familiar with RStudio. So perhaps this section should be moved after you've explained the panes and instead be more of an FYI for how to change the layout?

Author: I have had students accidentally change it and not realize it. Like they started with R years ago and changed their defaults and forgot :P and then they are confused when we go over the default

Reviewer: Fascinating. Okay then. Sounds good to leave it here then!

Click on the top menu of your RStudio - click where it says `Edit` --> `Preferences` --> `Pane Layout`.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Select Edit and then Preferences to get to the menu about pane layout."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g22370a5f292_430_0")
```

The Pane Layout menu enables you to change the layout. The image below shows the default settings. 

> ```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "The Pane Layout menu allows you to modify the layout of RStudio. This shows the default settings with Source on the top left, Console on the bottom left, Environment, History, Connections, and Build on the top right, and Files, Plots, Packages, Help, and VCS on the bottom right."}

Reviewer: What's VCS?

Author: a typo lol

Author: oh maybe I ment version control or something

Author: oh not a typo it is in the menu

Reviewer: I don't have VCS in my RStudio menu. Should we address that?

ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g22370a5f292_430_5")
```

</details>

> The pane on the left (labeled "Pane 1" in the image) is where we write code.

Reviewer: ```suggestion
The pane on the left (labeled "Pane 1" in the image) is where we can work on code interactively. There are two tabs here. The Terminal tab and the Console tab. The Terminal tab is for interacting with the computer outside of R. Whereas the Console tab is for interacting with R. We'll focus on the Console tab for now. 

The Console tab is where we can 'talk' to R and interactively work on our code. The code we write here will not be saved to a script or file, but instead the outcome of the code will immediately be performed we click `enter`.
```

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "The pane on the left is where we write and test code."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_81")
```

> The pane on the top right (labeled "Pane 2" in the image) is where we can see what objects we have created and are actively in memory (meaning they can be used at that time) in what is called the "Environment".

Reviewer: ```suggestion

Depending on what code we write in the Console tab, something will be echoed back to us or R will otherwise save it to its 'Environment'. The Environment in R can be thought of as R's working memory.

To play with the Console tab, let's practice using the assignment operator `<-`. The assignment operator is what creates R objects that are stored in the Environment. Aka the `<-` tells R to remember something for us. 

Try this bit of code: 
```
one_to_ten <- 1:10
```
If you look in the pane on the top right (labeled "Pane 2" in the image) we can see what objects we have created (and you should see `one_to_ten` there. The objects listed here are actively in R's memory or Environment. If we attempt to reference an object that is not in the Environment, R will give us an error. For example, if in the Console window we type `dog` R will say `Error: object 'dog' not found` because we have not created an object called `dog` yet.

The Console window is great for testing things out but note that for our code to be reproducible, we will need to eventually have it written down in a script or notebook because the Console window won't be something we can return to easily to rerun our code.

```

Author: Did you see the later example for this? I kinda just wanted this first part to orient people and then give them these examples, which is what I did next.

Author: I don't want to overwhelm them when they are first getting oriented

Reviewer: I generally try to have people walk through and apply things as soon as we introduce it so they don't have to hold information in their brain.

Reviewer: Where did we leave off with this bit?

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "The pane on the upper right is where we see our objects."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_95")
```

The pane on the bottom right (labeled "Pane 3" in the image) is where we can find files on computer (the "Files" tab), see plots (the "Plots tab), and get coding help (the "Help tab).

> ```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Whe pane on the lower right is where we see our files, plots, and where we get help."}

Reviewer: ```suggestion
```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "The pane on the lower right is where we see our files, plots, and where we get help."}
```

ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_109")
```

While there are other tabs, don't worry about those for now. We will go deeper into RStudio as we continue. 

> 

Reviewer: ```suggestion
<div class = "dictionary">
- **Console** - The window that allows us to interactively give R code and press enter to run it. 
- **Environment** - R's working memory of objects you have assigned -- told R to remember using `<-`
- **Assignment** - How we tell R to remember something using the `<-` characters. 
</div>
```

Let's try an example to get started.

As an example, we could type in the code `head(iris)` and press enter to see the code execute and preview.

> ```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Running some code in RStudio"}

Reviewer: ```suggestion
```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "In Pane one, in our Console window, we can run code interactively. Here we are running the code head(iris) which prints out a data set about flowers for us to look at. "}
```

Author: huh for me this slide seems to be about trying out code in the console

Reviewer: I wrote the wrong description it looks like? But I updated it now.

ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_142")
```

Now let's try another example where instead of just printing some data to the screen we assign a data object that will show up in the environment using the `<-` notation. This is useful in a situation if we want to modify the iris data somehow but want to keep the original version.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "If you create an object in RStudio, it will show up in the environment pane."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g1fa1583c827_0_4")
```

### The Hidden Pane

> There is a hidden fourth pane. This is only accessible when we start to make a script or a report with our code.

Reviewer: ```suggestion
There is a hidden fourth pane. This is only accessible when we start to make a script or a report with our code. 

In order to make our analysis truly reproducible we will need **every single step** written down. This is why using the Console is great for testing things, but not so great for actually performing your analysis. 

We are going to introduce you to a file that really helps make analyses reproducible: an R Markdown file. R Markdown files allow you to have the code for your analysis, the output from the analysis (so plots and stats, etc) and your written thoughts and rationale for your analysis all in one place! This makes it a snap to share your analysis with others in a reproducible way!
```

Reviewer: We should add a slide here of the parrot saying how great notebooks are because they have all these components.

Author: I have one of these in the next chapter do you think we should use it in both places? I give more context in the next chapter.

Author: I ended up splitting the text you suggested up and moving it around a tad but I added it.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Creating an R Markdown file in R Studio by selecting File, New File, R Markdown."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_127")
```

Creating an R Markdown file starts with a pop-up and you can simply click the OK button.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Creating an R Markdown file starts with a pop-up and you can simply click the OK button."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_169")
```

The new pane will open on the upper left.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "The new pane will open on the upper left."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_176")
```

This pane is where we can write code that we keep in files like scripts or reports (in files like R Markdowns).

> Thus the lower left pane is where we can test out code, but the top pane is where we can write code that we wish to save. 

Reviewer: ```suggestion
Thus the lower left pane is where we can test out code, but the top pane is where we can write code that we wish to save. 
```

I think we should try to move folks away from testing things in the console per se. Instead I thinking trying to have them highlight code in their script/notebook and run from there is better and more reproducible. People new to R have a tendency of writing everything in the console and underestimating how many steps they did outside their notebook that are not recorded and therefore not reproducible.

Author: I agree - but thought that the next chapter would really be focused on this, but I will add a statement here about that

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "The new pane will open on the upper left and is for writing code we want to save, while the bottom pane is for writing code we want to test."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed7630a13_1_186")
```

> The top pane where we save code is called the editor. The lower pane where we test code is called the console.

Reviewer: ```suggestion
The top pane where we save code is called the editor. The lower pane is the console which we explored previously. 
```
